### PR TITLE
rework  kube state metrics roles and collectors configuration

### PR DESCRIPTION
**Pull request by pgvishnuram**
## Description

This PR reworks the existing role and collector definition for kube state metrics and only generates roles for the enabled collectors 

## Related Issues

astronomer/issues#6389

## Testing

QA should now able to enable or disable or pass known list of collectors and should verify roles to see only those are created and applied

## Merging

cherry-pick to release-0.34 and release-0.35

### CODE REVIEW

File: charts/kube-state/values.yaml [hunk 1]
```diff
@@ -32,3 +32,35 @@ replicas: 1
 
 extraArgs:
 - "--metric-labels-allowlist=namespaces=[*],pods=[*]"
+
+priorityClassName: ~
+
+collectors:
```
Comment (rishkarajgi) [inline charts/kube-state/values.yaml:38]: TODO: Remove unused collectors and test metrics working without disruption

Comment (pgvishnuram) [reply charts/kube-state/values.yaml:38]: i have commented out the unused collectors

File: charts/kube-state/templates/kube-state-deployment.yaml [hunk 2]
```diff
@@ -28,6 +28,9 @@ spec:
         app: {{ template "kube-state.name" . }}
         version: {{ .Chart.Version }}
     spec:
+    {{- if .Values.priorityClassName }}
+      priorityClassName: {{ .Values.priorityClassName }}
+    {{- end }}
```
Comment (danielhoherd) [inline charts/kube-state/templates/kube-state-deployment.yaml:33]: There are no tests for this.

File: charts/kube-state/templates/kube-state-deployment.yaml [hunk 3]
```diff
@@ -42,6 +45,7 @@ spec:
       {{- if .Values.global.singleNamespace }}
         - --namespaces={{ .Release.Namespace }}
       {{- end }}
+        - --resources={{ .Values.collectors | join "," }}
```
Comment (danielhoherd) [inline charts/kube-state/templates/kube-state-deployment.yaml:48]: There are no tests for this.

File: tests/chart_tests/test_kube_state_deployment.py [hunk 4]
```diff
@@ -83,3 +83,30 @@ def test_kube_state_deployment_singleNamespace(self, kube_version):
...
+        assert len(docs[0]["rules"]) == 23
+
+    def test_kube_state_specific_collector_enabled(self, kube_version):
+        docs = render_chart(
+            kube_version=kube_version,
+            values={"kube-state": {"collectors": ["certificatesigningrequests"]}},
+            show_only=["charts/kube-state/templates/kube-state-role.yaml"],
+        )
+
+        assert len(docs[0]["rules"]) == 1
```
Comment (danielhoherd) [inline tests/chart_tests/test_kube_state_deployment.py:103]: We can do better here since we know that `certificatesigningrequests` will be in the rules.

**Review by danielhoherd — CHANGES_REQUESTED** [review]
I'd like to see better tests. I left some comments.

File: charts/kube-state/templates/kube-state-deployment.yaml (hunk 2 above)
Comment (pgvishnuram) [reply charts/kube-state/templates/kube-state-deployment.yaml:33]: test cases are added in here

File: tests/chart_tests/test_kube_state_deployment.py (hunk 4 above)
Comment (danielhoherd) [reply tests/chart_tests/test_kube_state_deployment.py:112]:
```suggestion
        assert len(docs[0]["rules"]) == 1
        assert "certificatesigningrequests" in docs[0]["rules"][0]["resources"]
```

**Review by danielhoherd — APPROVED** [review]
All of these changes look good to me. I think the feature is good and the reduction in prometheus metrics will be great. I left one optional suggestion to address the one remaining comment about improving a test case.

**Pull request merged**